DP-75 Publish to private repo

simonwgill opened the pull request:
Draft for the moment, since I want to see how it interacts with Jenkins.

@zhelyan I've added you as a reviewer so you can see what I'm trying to achieve so far.

simonwgill committed:
Fixed style issues in gem info

simonwgill committed:
Fixed unstable refs for consistent minor version

simonwgill committed:
Fixed style problems in Rakefile

simonwgill committed:
Updated config for private repo publishing

simonwgill committed:
Added prerelease number generation

simonwgill committed:
Added a mock build and package stage to check sense

simonwgill committed:
Fixed typos

simonwgill committed:
Moved scope of version variable

simonwgill committed:
Small fixes to jenkins variable interactions

simonwgill committed:
Fixed rubyism

simonwgill committed:
Testing vault integration

simonwgill committed:
Fixed typo

simonwgill committed:
Added missing dependency

simonwgill committed:
Changed echoes to executes

simonwgill committed:
Completed task from last commit

zhelyan commented:
I like the concept but I think we need to wrap it up in a standalone script / task dealing with the version bumping / naming. Otherwise you'll have to duplicate the logic in every repo and rely 100% on Jenkins to do the releases.
Tools like e.g.  https://github.com/svenfuchs/gem-release can provide a starting point.

Pushing the gems also affects the release process  so it would be useful to go back to it and answer some questions first:

- what should and should not be released. Are we publishing PR versions?
- what's involved in doing a release ( version bump, commits, changelog, pushing, notifications, readme? etc)
- who's doing the releases
- naming convention
- do we need pre-releases and if yes - how long should pre-releases stick around in the gem server
- where are we publishing
- how do we do a release
...

simonwgill commented:
A build on master would become a full release, and that uses the version number as it stands in the repository. If it's just that, it's pretty simple, because jenkins can call rake build, then the gem nexus call can just use pkg/*.gem

Prereleases is a weird one though - you've got to pick one of the main three components to bump when you package it, otherwise the actual release of that version gets picked up by dependendency management. That's why it's faking the release version through the rakefile. There's something of a problem with the way I've done this though - it won't push the prerelease artifact to nexus because the gemspec version and the filename version don't match.

I'd still like us to be able to use prerelease versions at least from the develop branch, but I'm still trying to figure out how to handle an incrementing build number for the prerelease identifier.

Agree on moving the logic into one place. Needs proving somewhere first though, which is why I picked this one.

I've added some pros and cons about github and nexus to https://citizensadvice.atlassian.net/wiki/spaces/OPS/pages/1671299091/Managing+releases+using+a+private+gem+server It turns out that it's easier to push and pull with nexus than github.

simonwgill committed:
Installed gem-release as a test

simonwgill committed:
Restricted behaviour to just pushing to nexus

Using env variables instead of interactive prompts

simonwgill committed:
Removed unnecessary gem install

simonwgill committed:
Added guard clauses with useful errors

simonwgill commented on Jenkinsfile:
```diff
@@ -45,5 +45,19 @@ node('docker && awsaccess') {
+
+      if (env.BRANCH_NAME == 'master') {
+        wrap([$class: 'VaultBuildWrapper', vaultSecrets: secrets]) {
+          sh("rake release")
```
The nexus repository currently allows overwriting of artifacts. Not 100% sure that's the right behaviour we want right now.

simonwgill commented:
@zhelyan I've stripped this back so it's just pushing artifacts when master gets merged and doing so by overriding the push subtask of the rake release task provided by bundler. Much cleaner, and doesn't have any weird version number shenaniganery in it.

zhelyan commented on Jenkinsfile:
```diff
@@ -45,5 +45,19 @@ node('docker && awsaccess') {
     }
+
+    stage('Push to repository') {
+      def secrets = [
```
just a personal preference - you can use the much more concise https://github.com/citizensadvice/jenkins-global-pipeline-lib/blob/master/vars/withVaultSecrets.txt

simonwgill replied:
I'd forgotten about that. Easy to fix.

Any other issues before I widen out the reviews?

zhelyan replied:
looks good to me. The only thing I would try is to leverage some of the internal methods here https://github.com/rubygems/bundler/blob/35be6d9a603084f719fec4f4028c18860def07f6/lib/bundler/gem_helper.rb#L98

e.g. `sh_with_input`

zhelyan replied:
..which is protected. :-))

simonwgill committed:
Used existing vault secrets function

(includes check that Jenkins is behaving correctly this will be removed
before merge)

simonwgill committed:
Fixed type

simonwgill committed:
Removed testing output

simonwgill committed:
Removed extraneous require

ruthwells reviewed (approved):
We talked through it - it all seems reasonable, even if I couldn't reproduce it from scratch!
